### PR TITLE
AnimatedList sample app

### PR DESCRIPTION
**Pull request by HansMuller**
A sample app that demonstrates how to use AnimatedList. It includes a ListModel class, that demos how to link a list data model with the AnimatedList.

The app is lightly documented. I hope that in the near future:
- The AnimatedList dart doc will link to it
- The widget catalog will include an entry for AnimatedList and link to it

In the slightly longer term I hope that:
- We'll automatically generate a screenshot of the app in action
- The widget catalog will incorporate the screenshot
- There will be more samples 

This is all `flutter create` generated boilerplate except for:

examples/samples/animated_list/pubspec.yaml
examples/samples/animated_list/lib/main.dart
examples/samples/animated_list/test/smoke_test.dart

### CODE REVIEW

**Comment by abarth** [comment]
What does the "samples" part mean?  I was expecting this to just be in examples/animated_list

**Comment by Hixie** [comment]
Maybe `examples/widgets/*` ?

@abarth The idea is to build a little example app for each widget, as I understand it.

**Comment by abarth** [comment]
I'm worried that gradle/Xcode overhead will make that prohibitive.  Can we have one sample app that shows off each in a separate page?  That sounds a bit like flutter_gallery...

**Comment by HansMuller** [comment]
The goal is for each sample to be a self-contained app. There doesn't really need to be one sample app per flutter project, for example we could have a collections of samples per TBD widget category (I don't think a flutter gallery style viewer for all of the samples is needed either). That said, perhaps we should add directory substructure as needed. How about I just move this first sample up a level and move lib/main.dart to lib/animated_list_sample.dart?

**Comment by Hixie** [comment]
We could do this layers-style. The layers directory has one "app" but many "main.dart" files (named various things).

**Comment by abarth** [comment]
Yeah, maybe we should merge layers into this example at some point.

**Comment by HansMuller** [comment]
I've set things up so that many sample apps can be moved in examples/samples/lib.

I'll add some scaffold/appbar samples next to help flesh out the way things are organized.

File: examples/catalog/lib/animated_list.dart [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+/// Keeps a Dart List in sync with an AnimatedList.
+///
+/// The [insert] and [removeAt] methods apply to both the internal list and the
+/// animated list that belongs to [listKey].
+///
+/// This class only exposes as much of the Dart List API as is needed by the
+/// sample app. More list methods are easily added, however methods that mutate the
+/// list must make the same changes to the animated list in terms of
+/// [AnimatedListState.insertItem] and [AnimatedList.removeItem].
+class ListModel<E> {
```
Comment (Hixie) [inline examples/catalog/lib/animated_list.dart]: i believe there's a mixin you can use to make this trivial.
See https://docs.flutter.io/flutter/dart.collection/ListMixin-class.html

**Comment by Hixie** [comment]
The example looks a bit more complicated than I would have expected from such a thing, but LGTM.

File: examples/catalog/lib/animated_list.dart (hunk 1 above)
Comment (HansMuller) [inline examples/catalog/lib/animated_list.dart]: That's an excellent idea.  I will update this in a separate PR.

**Comment by diogoribeiro23** [comment]
Hey! Sorry for my question but why the internal list? I altered the sample to only use the animated list (and its state) and it works fine.

**Pull request merged**